### PR TITLE
Add support for choosing a process to reply to (new)

### PR DESCRIPTION
**Pull request by manifest**
Related to the [PR](https://github.com/ninenines/gun/pull/109) and the [issue](https://github.com/ninenines/gun/issues/108).

### CODE REVIEW

**Comment by essen** [comment]
Thanks, it's running. I'll ping you if there's some issues.

**Comment by essen** [comment]
No problem with the tests, just needs to find time to work on it now. Thanks!

**Comment by essen** [comment]
Can you update the documentation to mention a possible issue with Websocket I describe here? https://github.com/ninenines/gun/issues/108#issuecomment-260956989

**Comment by manifest** [comment]
I've updated the guide and the manual.

**Comment by essen** [comment]
Nice work. I'll get back to you.

File: doc/src/guide/http.asciidoc [hunk 1]
```diff
@@ -222,10 +222,24 @@ gun:request(ConnPid, "TRACE", "/", [
...
+
+[source,erlang]
+ReqOpts = #{reply_to => HandlerPid},
+StreamRef = gun:get(ConnPid, "/organizations/ninenines", [], ReqOpts).
+
 === Processing responses
 
-All data received from the server is sent to the owner
-process as a message. First a `gun_response` message is sent,
+First a `gun_response` message is sent,
```
Comment (essen) [inline doc/src/guide/http.asciidoc]: I would keep this as-is, and put the `Redirecting responses to a different process` part as its own section at the end of the document instead. It's a special case, most people won't care.

File: doc/src/guide/websocket.asciidoc [hunk 2]
```diff
@@ -64,6 +64,11 @@ after 1000 ->
 	exit(timeout)
 end.
 
+Note that you shouldn't use `reply_to` request option
+for connections you plan to upgrade. By doing it,
```
Comment (essen) [inline doc/src/guide/websocket.asciidoc]: By doing so

**Comment by essen** [comment]
Clicked review by mistake, sorry about that.

File: doc/src/manual/gun.asciidoc [hunk 3]
```diff
@@ -241,39 +252,54 @@ effectively closed.
 
 === delete(ConnPid, Path) -> delete(ConnPid, Path, [])
 
-Alias of `gun:delete/3`.
+Alias of `gun:delete/4`.
 
 === delete(ConnPid, Path, Headers) -> StreamRef
```
Comment (essen) [inline doc/src/manual/gun.asciidoc]: You need to update this part also (see delete/2).

Comment (essen) [inline doc/src/manual/gun.asciidoc]: Same goes for the other functions.

File: src/gun.erl [hunk 4]
```diff
@@ -268,23 +305,35 @@ put(ServerPid, Path, Headers) ->
...
 
 -spec data(pid(), reference(), fin | nofin, iodata()) -> ok.
 data(ServerPid, StreamRef, IsFin, Data) ->
-	_ = ServerPid ! {data, self(), StreamRef, IsFin, Data},
+	data(ServerPid, StreamRef, IsFin, Data, #{}).
+
+-spec data(pid(), reference(), fin | nofin, iodata(), req_opts()) -> ok.
+data(ServerPid, StreamRef, IsFin, Data, ReqOpts) ->
+	ReplyTo = maps:get(reply_to, ReqOpts, self()),
+	_ = ServerPid ! {data, ReplyTo, StreamRef, IsFin, Data},
```
Comment (essen) [inline src/gun.erl]: Requests I understand, but data/cancel not sure I see the use case.

Comment (manifest) [inline src/gun.erl]: I guess that sending an [error message](https://github.com/manifest/gun/blob/feature/reply-to/src/gun_http.erl#L297) to the process that calls `data` or `cancel` function would be an expected behaviour.

Comment (essen) [inline src/gun.erl]: Sure but what about the process that initialized the request? I can understand different processes for different requests, but many processes for one request bends my brain a bit too much.

Comment (manifest) [inline src/gun.erl]: Yeah, I don't see any real use case here. I will remove the `ReqOpts` argument for `data` and `cancel` functions.

**Comment by manifest** [comment]
Ready

**Comment by essen** [comment]
Cheers. Probably on Thursday next for me. :-)

**Comment by essen** [comment]
I'm too busy right now. I'll come back later, sorry.

**Comment by manifest** [comment]
@essen I've just wanted to remind you about this one

**Comment by essen** [comment]
I'm working on this right now FYI. Making a few changes of things I didn't notice before, then will be pushed.

**Comment by essen** [comment]
OK I'm pushing. I still have mixed feelings about a few things but it can be addressed at a later time. Thanks!

**Pull request closed without merging**